pkg/redact/redact.go

```diff
@@ -48,6 +49,36 @@ func Redact(msg interface{}, val reflect.Value) {
 	}
 }
 
+// Redact - removes data from fields marked as sensitive given a proto message struct
+func RedactSensitiveError(msg interface{}, val reflect.Value, err *error) {
+	// TODO: This needs to be optimized!
+	// Should cache messages that contain no sensitive data to ignore, and should cache the map of tag number to field name
+
+	if !val.IsValid() {
+		return
+	}
+
+	switch val.Kind() {
+	case reflect.Slice:
+		for i := 0; i < val.Len(); i += 1 {
+			RedactSensitiveError(val.Index(i).Interface(), val.Index(i), err)
+		}
+	case reflect.Map:
+		// TODO: Implement map logic. Currently only certificates in identity use it, and are redacted
+	case reflect.Ptr:
+		RedactSensitiveError(msg, val.Elem(), err)
+	case reflect.Struct:
+		redactErrorMessage(msg, val, err)
+	}
+}
+
+// RedactedMessage - returns a copy of a proto message struct with data from fields marked as sensitive redacted
+func RedactedError(msg interface{}, err *error) interface{} {
+	rMsg := proto.Clone((msg).(proto.Message))
+	RedactSensitiveError(rMsg, reflect.ValueOf(rMsg), err)
+	return rMsg
+}
+
 func redactMessage(msg interface{}, val reflect.Value) {
```

Comment:
NIT: Although we did quite a bit of code refactoring, it only applies to redacting sensitive information from error messages. The implementation is very similar to redacting sensitive value from the field, but we have duplicated code. We should be able to reuse it with some form of default values, such as:

RedactMessageWithDefault(msg) {
    RedactMessage(msg, nil)
}
Given the urgency, we could just file an item to track future improvements for refactoring.

Reply:
Will add future work item for it